nexus/types/src/external_api/params.rs

```diff
@@ -2741,6 +2778,301 @@ pub struct AuditLog {
+/// Path parameters for instance multicast group operations.
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema)]
+pub struct InstanceMulticastGroupPath {
+    /// Name or ID of the instance
+    pub instance: NameOrId,
+    /// Name or ID of the multicast group
+    pub multicast_group: NameOrId,
+}
+
+/// Validate that an IP address is suitable for use as a SSM source.
+///
+/// For specifics, follow-up on RFC 4607:
+/// <https://www.rfc-editor.org/rfc/rfc4607>
+pub fn validate_source_ip(ip: IpAddr) -> Result<(), String> {
+    match ip {
+        IpAddr::V4(ipv4) => validate_ipv4_source(ipv4),
+        IpAddr::V6(ipv6) => validate_ipv6_source(ipv6),
+    }
+}
+
+/// Validate that an IPv4 address is suitable for use as a multicast source.
+fn validate_ipv4_source(addr: Ipv4Addr) -> Result<(), String> {
+    // Must be a unicast address
+    if !is_unicast_v4(&addr) {
+        return Err(format!("{} is not a unicast address", addr));
+    }
+
+    // Exclude problematic addresses (mostly align with Dendrite, but block link-local)
+    if addr.is_loopback()
+        || addr.is_broadcast()
+        || addr.is_unspecified()
+        || addr.is_link_local()
+    {
+        return Err(format!("{} is a special-use address", addr));
+    }
+
+    Ok(())
+}
+
+/// Validate that an IPv6 address is suitable for use as a multicast source.
+fn validate_ipv6_source(addr: Ipv6Addr) -> Result<(), String> {
+    // Must be a unicast address
+    if !is_unicast_v6(&addr) {
+        return Err(format!("{} is not a unicast address", addr));
+    }
+
+    // Exclude problematic addresses (align with Dendrite validation, but block link-local)
+    if addr.is_loopback()
+        || addr.is_unspecified()
+        || ((addr.segments()[0] & 0xffc0) == 0xfe80)
+    // fe80::/10 link-local
+    {
+        return Err(format!("{} is a special-use address", addr));
+    }
+
+    Ok(())
+}
+
+/// Validate that an IP address is a proper multicast address for API validation.
+pub fn validate_multicast_ip(ip: IpAddr) -> Result<(), String> {
+    match ip {
+        IpAddr::V4(ipv4) => validate_ipv4_multicast(ipv4),
+        IpAddr::V6(ipv6) => validate_ipv6_multicast(ipv6),
+    }
+}
+
+/// Validates IPv4 multicast addresses.
+fn validate_ipv4_multicast(addr: Ipv4Addr) -> Result<(), String> {
+    // Verify this is actually a multicast address
+    if !addr.is_multicast() {
+        return Err(format!("{} is not a multicast address", addr));
+    }
+
+    // Define reserved IPv4 multicast subnets using oxnet
+    //
+    // TODO: Eventually move to `is_reserved` possibly?...
+    // https://github.com/rust-lang/rust/issues/27709
+    let reserved_subnets = [
+        // Local network control block (link-local)
+        Ipv4Net::new(Ipv4Addr::new(224, 0, 0, 0), 24).unwrap(),
+        // GLOP addressing
+        Ipv4Net::new(Ipv4Addr::new(233, 0, 0, 0), 8).unwrap(),
+        // Administrative scoped addresses
+        Ipv4Net::new(Ipv4Addr::new(239, 0, 0, 0), 8).unwrap(),
```

Comment:
Do we really want to block admin scope? I could see this being used for an on-prem deployment.